### ISSUE TITLE
Live stream progress updates from backend to frontend

### ISSUE THREAD

**Issue by jonc125** [UCL-CloudLabs/user-portal#7]
So when a resource is being created, users have the option of looking at a pane showing the command output as it happens, like for Travis builds etc.

Sub-tasks to make this happen:
- [x] Use e.g. [Celery](http://flask.pocoo.org/docs/0.12/patterns/celery/) as a task queue to do deployment in the background, so the browser gets an initial response quickly. (Note: the [Flask-Celery](https://pypi.python.org/pypi/Flask-Celery) package should not be used as of Celery 3.)
    - [x] Figure out how to handle output from this background task - Celery tasks shouldn't normally create lots of output; we may want to store it in the DB.
    - [ ] The page returned when you add a host should no longer be the dashboard, but a new page saying that deployment is in progress.
- [ ] Add a REST API view that retrieves the latest output from the background task.  (Which should be fairly trivial if it's in the DB, and probably nasty otherwise!)
- [ ] Add JS code in the 'in progress' page above, which regularly polls the above API and appends the latest output to a box.
- [ ] Handle failures gracefully (#15)

**Comment by jonc125** [comment]
See also https://github.com/UCL-CloudLabs/deployer/issues/13 which is a prerequisite for this.

**Comment by jonc125** [comment]
See also https://github.com/UCL-CloudLabs/deployer/issues/3.

**Comment by raquelalegre** [comment]
Still working on prerequisite https://github.com/UCL-CloudLabs/deployer/issues/13

**Comment by jonc125** [comment]
Have started looking into Celery in the celery branch, and thinking about how to approach it. Some ideas:
- [x] Change the status column to an `Enum`.  Not strictly required, but more rigorous than checking against string values! We'll need to tweak an alembic setting so the auto-migration notices this I think.
- [x] Add an 'error' status value, used if deployment failed for some reason, along with a 'deploying' status while it's happening. (Value added, but not used yet.)
- [x] Add a `deploy_log` column to the `Host` table for recording Terraform etc. output
- [x] When we call deploy, this launches a Celery task passing it the host id only. The front-end returns immediately to the Dashboard, showing the status as 'deploying'. (Or possibly goes to the deploy log view.)
    - [x] The Celery task handles calling Terraform, and updates the `deploy_log` column each time there's output.
    - (If it notices the row has been deleted, this means the user has deleted the host, so it should abort the deployment.)
    - [x] There should be a new view, reachable for deploying hosts (or ones that failed to deploy), that displays the `deploy_log` contents in a box. It should (initially) auto-refresh until the status is no longer 'deploying'.
    - [ ] Later we can use some JS to refresh just the log portion of the page, calling a REST view to get that text.

**Comment by jonc125** [comment]
SocketIO (https://flask-socketio.readthedocs.io/en/latest/) might be useful if we want the server to push updates to the client eventually.

**Comment by jonc125** [comment]
We should check that calls like `host.update(deploy_log=host.deploy_log + line.decode('utf-8'))` generate a concat within the database, rather than reading the existing value, concat in Python, then write back. Need to figure out how to view the SQL that's generated.

**Comment by jonc125** [comment]
Seems the reason deploying wasn't working on my system is that I hadn't set the environment variables for Terraform, so it was waiting for them to be provided on stdin. I discovered this when I killed the celery worker with Ctrl-C, which caused Terraform to do a graceful shutdown and print extra output:
```
var.azure_client_id
  Enter a value: 
Interrupt received.
Please wait for Terraform to exit or data loss may occur.
Gracefully shutting down...
```

The unfortunate part of this is that Terraform seems to return a 0 exit code when this happens, and the host's status became 'running'! We'll have to check this, especially in the context of #15.
